tests/TransactionTest.php

```diff
@@ -0,0 +1,430 @@
+
+            // Update user during transaction to simulate a simultaneous update
+            if ($timesRun == 0) {
+                DB::getCollection('users')->updateOne(['name' => 'klinson'], ['$set' => ['age' => 22]]);
+            }
+
+            $timesRun++;
+        }, 2);
+
+        $this->assertSame(2, $timesRun);
+        $this->assertTrue(User::where(['name' => 'klinson'])->where('age', 21)->exists());
+    }
+
+    public function testTransactionRespectsRepetitionLimit(): void
+    {
+        $klinson = User::create(['name' => 'klinson', 'age' => 20, 'title' => 'admin']);
+
+        $timesRun = 0;
+
+        DB::transaction(function () use (&$timesRun): void {
+            User::where(['name' => 'klinson'])->update(['age' => 21]);
+
+            // Update user during transaction to simulate a simultaneous update
+            DB::getCollection('users')->updateOne(['name' => 'klinson'], ['$inc' => ['age' => 2]]);
+
+            $timesRun++;
+        }, 2);
+
+        $this->assertSame(2, $timesRun);
+
+        $check = User::find($klinson->_id);
+        $this->assertInstanceOf(User::class, $check);
+
+        // Age is expected to be 24: the callback is executed twice, incrementing age by 2 every time
+        $this->assertSame(24, $check->age);
+    }
+
+    public function testTransactionReturnsCallbackResult(): void
+    {
+        $result = DB::transaction(function (): User {
+            return User::create(['name' => 'klinson', 'age' => 20, 'title' => 'admin']);
+        });
+
+        $this->assertInstanceOf(User::class, $result);
+        $this->assertEquals($result->title, 'admin');
+        $this->assertSame(1, User::count());
+    }
+
+    public function testNestedTransactionsCauseException(): void
+    {
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Transaction already in progress');
+
+        DB::beginTransaction();
+        DB::beginTransaction();
+        DB::commit();
+        DB::rollBack();
+    }
+
+    public function testNestingTransactionInManualTransaction()
+    {
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('Transaction already in progress');
+
+        DB::beginTransaction();
+        DB::transaction(function (): void {
+        });
+        DB::rollBack();
+    }
+
+    public function testCommitWithoutSession(): void
+    {
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('There is no active session.');
+
+        DB::commit();
+    }
+
+    public function testRollBackWithoutSession(): void
+    {
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('There is no active session.');
+
+        DB::rollback();
+    }
+
+    private function getPrimaryServerType(): int
+    {
+        return DB::getMongoClient()->getManager()->selectServer(new ReadPreference(ReadPreference::RP_PRIMARY))->getType();
```

Comment:
`composer.json` currently requires `mongodb/mongodb: ^1.11`, which requires `ext-mongodb: ^1.10`. If you'd like to make this more concise, you can bump the PHPLIB requirement to `^1.12` and call [`selectServer()`](https://www.php.net/manual/en/mongodb-driver-manager.selectserver.php) without any parameter for the same effect.

Reply:
It's a good thing that we've decided to align the version numbers of the library and the extension. The change to `selectServer()` was made in `ext-mongodb` 1.11, which is required as of `mongodb/mongodb` 1.10. Somehow the fact that we default to a primary read preference had escaped my memory. I've updated the call.